File: Java/74.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0) return false;
        int m=matrix.length, n=matrix[0].length;
        int i=0, j= m*n-1;
        while (i<=j) {
            int mid = i + (j-i)/2;
            int x = mid / n;
            int y = mid % n;
            if (matrix[x][y] == target) return true;
            if (matrix[x][y] < target) {
                i = mid + 1;
            } else {
                j = mid - 1;                
            }
        }
        return false;
    }
}
__________________________________________________________________________________________________
sample 35376 kb submission
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int low;
        int high;
        int mid;
        if(matrix.length == 0) return false;
        for(int[] arr : matrix){
            if(arr.length == 0) continue;
            if(arr.length == 1 && arr[0] == target) return true;
            if(arr.length > 0 && target <= arr[arr.length-1]){
                 low = 0;
                 high = arr.length - 1;
                while(low <= high){
                    mid = (high - low) + low / 2;
                    if(arr[mid] < target)
                        low = mid + 1;
                    else if (arr[mid] > target)
                        high = mid - 1;
                    else if(arr[mid] == target)
                        return true;                
                }
                return false;
            }
          }
        return false;
    }
}
__________________________________________________________________________________________________

File: Python3/74.py
__________________________________________________________________________________________________
sample 64 ms submission
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#
# https://leetcode.com/problems/search-a-2d-matrix/description/
#
# algorithms
# Medium (34.42%)
# Total Accepted:    186.8K
# Total Submissions: 542.8K
# Testcase Example:  '[[1,3,5,7],[10,11,16,20],[23,30,34,50]]\n3'
#
# Write an efficient algorithm that searches for a value in an m x n matrix.
# This matrix has the following properties:
# 
# 
# Integers in each row are sorted from left to right.
# The first integer of each row is greater than the last integer of the
# previous row.
# 
# 
# Example 1:
# 
# 
# Input:
# matrix = [
# ⁠ [1,   3,  5,  7],
# ⁠ [10, 11, 16, 20],
# ⁠ [23, 30, 34, 50]
# ]
# target = 3
# Output: true
# 
# 
# Example 2:
# 
# 
# Input:
# matrix = [
# ⁠ [1,   3,  5,  7],
# ⁠ [10, 11, 16, 20],
# ⁠ [23, 30, 34, 50]
# ]
# target = 13
# Output: false
# 
#
class Solution:
    def searchMatrix(self, matrix, target):
        if not matrix or not matrix[0]:
            return False
        r, c = len(matrix), len(matrix[0])
        f, l = 0, r * c
        while f < l:
            m = f + (l - f) // 2
            if matrix[m // c][m % c] < target:
                f = m + 1
            else:
                l = m
        return f < r * c and matrix[f // c][f % c] == target        
__________________________________________________________________________________________________
sample 13712 kb submission
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def isListEmpty(inList):
            if isinstance(inList, list): # Is a list
                return all( map(isListEmpty, inList) )
            return False
        if isListEmpty(matrix):
            return False
        numrows = len(matrix)
        numcols = len(matrix[0])
        col = 0
        row = 0
        while row < numrows:
            if target == matrix[row][col]:
                return True
            elif target > matrix[row][col] and row != (numrows -1):
                row += 1
            elif target > matrix[row][col] and row == (numrows -1):
                break
            elif target < matrix[row][col] and row != 0:
                row -= 1
                break
            else:
                break
        while col < numcols:
            if target == matrix[row][col]:
                return True
            elif target > matrix[row][col] and col != (numcols -1):
                col += 1
            elif target > matrix[row][col] and col == (numcols -1):
                return False
            elif target < matrix[row][col] and col != 0:          
                return False
            else:
                return False
__________________________________________________________________________________________________
